### DISCUSSION TITLE
Filtering account retrieval

### DISCUSSION THREAD

**Discussion by janko** [jeremyevans/rodauth#267] [General]
I'm currently [recommending](https://github.com/janko/rodauth-rails/wiki/Account-Types) users of rodauth-rails to share the same `account*` tables between Rodauth configurations, to keep the database schema DRY, and store an account type identifier in the database.

The account type is saved on account creation, and retrieved to filter the account. I first thought I only need to override the `account_ds` method, but I realized during the live stream that `_account_from_login` would need to be overridden as well. Below is a full example of such setup:

```rb
require "rodauth"
require "roda"
require "sequel"

DB = Sequel.sqlite
DB.create_table :accounts do
  primary_key :id
  Integer :status_id, null: false, default: 1
  String :email, null: false
  String :password_hash
  String :type, null: false
end

class RodauthBase < Rodauth::Auth
  configure do
    enable :create_account, :login, :internal_request
    account_password_hash_column :password_hash
    before_create_account { account[:type] = account_type }
    account_from_login do |login|
      account = super(login)
      account if account[:type] == account_type
    end
  end

  private

  def account_ds(*)
    super.where(type: account_type)
  end

  def account_type
    self.class.configuration_name&.to_s || "main"
  end
end

class RodauthMain < RodauthBase
end

class RodauthAdmin < RodauthBase
end

class RodauthApp < Roda
  plugin :rodauth, auth_class: RodauthMain
  plugin :rodauth, auth_class: RodauthAdmin, name: :admin
end

RodauthMain.create_account(login: "user@example.com", password: "secret")
RodauthMain.account_exists?(login: "user@example.com") # => true
RodauthAdmin.account_exists?(login: "user@example.com") # => false
```

If this is legit from the design perspective, my question is whether it's worth making the account filtering easier, so that it only needs to be done in one place instead of two (`account_ds` and `_account_from_login`). I was thinking of extracting `db[accounts_table]` in a separate method, but I don't know how to call it. Another idea was adding an `account_filter` configuration method, which would be passed to `where` in both `account_ds` and `_account_from_login`, and would default to an empty hash. What do you think?

**Comment by jeremyevans** [comment]
My recommendation would be to use separate tables for separate types of users.  However, I would be OK adding an `account_table_ds` private method that both `account_ds` and `account_from_login` use.  I don't want to add an `account_filter` method.

**Reply by janko** [reply]
Great, thanks, I will send a PR then. And I will include the option to use separate tables in the wiki page 👍
